src/Listener/JsonApiListener.php

```diff
@@ -0,0 +1,281 @@
+            throw new CrudException('JsonApiListener configuration option `include` only accepts an array');
+        }
+        $controller->set([
+            '_include' => $this->config('include')
+        ]);
+
+        if (!is_array($this->config('fieldSets'))) {
+            throw new CrudException('JsonApiListener configuration option `fieldSets` only accepts an array');
+        }
+        $controller->set([
+            '_fieldSets' => $this->config('fieldSets')
+        ]);
+
+        // Set required viewVar with ORM AssociationCollection for the current
+        // model so it can be used to generate the `relationships` nodes in the
+        // response.
+        //
+        // Please note that we are removing associated models not found in
+        // the find() result to prevent `null` relationships appearing in
+        // the response.
+        $controllerName = $controller->name; // e.g. Countries
+        $entityName = Inflector::singularize($controllerName); // e.g. Country
+        $table = $controller->$controllerName; // table object
+
+        if ($controller->request->action === 'index') {
+            $findResult = $subject->entities;
+            $firstEntity = $subject->entities->first();
+        } else {
+            $findResult = $subject->entity;
+            $firstEntity = $subject->entity;
+        }
+
+        $associations = $table->associations();
+
+        foreach ($associations as $association) {
+            $associationKey = strtolower($association->name());
+            $entityKey = $association->table();
+
+            if (get_class($association) === 'Cake\ORM\Association\BelongsTo') {
+                $entityKey = Inflector::singularize($entityKey);
+            }
+
+            if (empty($firstEntity->$entityKey)) {
+                $associations->remove($associationKey);
+            }
+        }
+
+        $controller->set(['_associations' => $associations]);
+
+        // Set required viewVar with array holding entity names of current
+        // entity and all related/contained models. Used to generate/read
+        // NeoMerx schemas inside the view.
+        $entities = [$entityName];
+
+        foreach ($associations as $association) {
+            $entities[] = Inflector::singularize($association->name());
+        }
+
+        $controller->set(['_entities' => $entities]);
+
+        // In debug mode set queryLog viewVar
+        if (Configure::read('debug')) {
+            $controller->set([
+                '_queryLog' => $this->_getQueryLog()
+            ]);
+        }
+
+        // Set data before rendering the view
+        $controller->set([
+            Inflector::tableize($controller->name) => $findResult,
+            '_serialize' => true,
+        ]);
+
+        return $controller->render();
+    }
+
+    /**
+     * Helper method to get query log.
+     *
+     * @return array Query log.
+     */
+    protected function _getQueryLog()
```

Comment:
You are totally right, I did not want to touch (or depend on) anything existing but your suggestion solves all that so I I just moved this into a trait in https://github.com/FriendsOfCake/crud/pull/445/files#diff-62d4f5f8797dc1fd34406920bd2a38baR164. Thanks, much obliged 💃